### ISSUE TITLE
Create a wiki page about best practices for maintaining GitHub secrets

### ISSUE THREAD

**Issue by SAUMILDHANKAR** [hackforla/website#3290]
### Overview
As a developer, I would like all the secrets being used in the website repo to be well maintained. In this issue, we will create a wiki page about the best practices for maintaining and resetting secrets in GitHub repos.

### Action Items
- [x] We are currently in the process of moving the old wiki to the new [website-wiki repo](https://github.com/hackforla/website-wiki), so we will not be making any changes or additions to the old wiki at this time. Thus, we will be adding wiki content through a different process now. Read [How to Contribute to the Wiki](https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki)
- [x] Explore how [GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) and tokens work.
- [x] Research the best practices for resetting secrets in general, including the `Audit and rotate registered secrets` as well as `Audit how secrets are handled` sections in this helpful article on [using secrets](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-secrets).
- [x] Decide on what would be the best place to add your research to [HfLA wiki pages](https://github.com/hackforla/website/wiki)
- [x] Create a draft wiki page in a comment in this issue, and list your recommendations/suggestions that might be helpful for the website team going forward.  
- [x] Once done, place this issue in the `Questions/In Review` column and add the `ready for dev lead` label. 
- [x] When the draft is approved, add a link to the draft to [How to Contribute to the Wiki](https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki)
- [x] **For reviewers**: Once this issue is approved and closed, check off the dependency in #3228, remove the dependency label and move issue #3228 from ice box column to the new issue approval column.

### Resources/Instructions
- [GitHub encrypted secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets)
- [HfLA wiki pages](https://github.com/hackforla/website/wiki)
- Note from a prior assignee to this issue:
>I did speak with some open source repository maintainers and they said that they just waited for the emails. They suggested that we make sure that the emails from github are visible. I think we have a problem with someone being responsible for checking the email that the repository sends to. We might consider adding a notification to the channel or calendar to update secrets. It is not considered best practice to never update secrets.

**Comment by ExperimentsInHonesty** [comment]
@SAUMILDHANKAR There will need to be an action item to add something to the wiki so that people in the future can find the document.  Also, as much as possible should be added to the wiki except where it compromises security.

**Comment by SAUMILDHANKAR** [comment]
@ExperimentsInHonesty I have added an action item to add the research information to a wiki page. This issue won't have any secure information related to our repo so all of the research can go on the wiki page. Thanks.

**Comment by JessicaLucindaCheng** [comment]
@SAUMILDHANKAR Is this ready for a "ready for milestone" label? If yes, please add the "ready for milestone" label and remove the "ready for dev lead" label. Thanks.

**Comment by blulady** [comment]
availability: 22-9-23/26
eta: 22-9-29

**Comment by blulady** [comment]
Progress update:
got side tracked 
Availability: 22-10-1/2
ETA: 22-10-3

**Comment by blulady** [comment]
Progress: sitting down with this now
Blockers: just haven't had time
Availability: 6 hours this week
Eta: unknown
Pictures: None, probably won't be any, we'll see

**Comment by blulady** [comment]
Progress: have searched the internet for best practices, did speak with a dev about how they handle theirs -very project specific, recommends changing often especially if we have logs because they can print out the secrets
Blockers: haven't found a good protocol to apply, need to talk it over with the team
Availability: 6 hours this week
Eta: unknown
Pictures: None, probably won't be any, we'll see

**Comment by blulady** [comment]
Progress: Have spoken to mentor/other Developers about GH Secrets maintenance was advised to change regularly and that often times they can be printed out by logs
Blockers: Other more pressing concerns have prevented me from having a discussion with the team about how we want to handle this
Availability: a few hours on Monday
ETA: After I speak with everyone else, I am looking forward to making a formal write up

**Comment by blulady** [comment]
Talking to Devs this weekend.

**Comment by ExperimentsInHonesty** [comment]
@blulady If you are not working on this anymore, can you finish making your notes and stick it back in the backlog?

**Comment by Adaakal** [comment]
@blulady

Please add update using the below template (even if you have a pull request). Afterwards, remove the 'To Update !' label and add the 'Status: Updated' label.

Progress: "What is the current status of your project? What have you completed and what is left to do?"
Blockers: "Difficulties or errors encountered."
Availability: "How much time will you have this week to work on this issue?"
ETA: "When do you expect this issue to be completed?"
Pictures (optional): "Add any pictures of the visual changes made to the site so far."
If you need help, be sure to either: 1) place your issue in the developer meeting discussion column and ask for help at your next meeting, 2) put a "Status: Help Wanted" label on your issue and pull request, or 3) put up a request for assistance on the #hfla-site channel.

**Comment by blulady** [comment]
Progress: I need to reach out to other projects to see how they manage their github secrets
    Blockers: Have not found much useful documentation online
    Availability: 2 hours
    ETA: 3/3

**Comment by JessicaLucindaCheng** [comment]
@blulady We are currently in the process of moving the old wiki to the new [website-wiki repo](https://github.com/hackforla/website-wiki), so we will not be making any changes or additions to the old wiki at this time. Thus, we will be adding wiki content through a different process now. Read [How to Contribute to the Wiki](https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki)

**Comment by ExperimentsInHonesty** [comment]
@blulady how is it going on this issue?  What information are you looking for online that you are not finding?

**Comment by blulady** [comment]
> @blulady how is it going on this issue? What information are you looking for online that you are not finding?

Hey Bonnie, I am not finding the information online. I am reaching out now to some open source maintainers to ask how they handle it.

**Comment by ExperimentsInHonesty** [comment]
@blulady is this what you are looking for? https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository. 

if its not, can you explain what you are looking for?

**Comment by blulady** [comment]
@ExperimentsInHonesty Sorry!! I thought someone took me off of this already. I did speak with some open source repository maintainers and they said that they just waited for the emails. They suggested that we make sure that the emails from github are visible. I think we have a problem with someone being responsible for checking the email that the repository sends to. We might consider adding a notification to the channel or calendar to update secrets. It is not considered best practice to never update secrets. I am not sure where to go from here.
Progress: Spoke with repository maintainers, have a general idea of what we could do to maintain secrets.
Blockers: Need to speak with the team to see how they want to proceed.
Availability: 1 hour this week
ETA: after speaking with the team.

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty I don't think it is necessary to restrict this issue to dev leads.  I suggest we can reomove the `role: dev leads` and add `role: back end/devOps`.  Then it is ready for prioritization.

**Comment by HackforLABot** [comment]
Hello @k-cardon, we appreciate you taking on this issue, however it looks like you're already working on another issue at this time. Please wait until your current issue is merged before taking on another issue. :)

We are going to unassign you from this issue so you can focus on your current issue.

Hfla appreciates you! :)

**Comment by HackforLABot** [comment]
Hello @k-cardon, we appreciate you taking on this issue, however it looks like you're already working on another issue at this time. Please wait until your current issue is merged before taking on another issue. :)

We are going to unassign you from this issue so you can focus on your current issue.

Hfla appreciates you! :)

**Comment by HackforLABot** [comment]
Hi @k-cardon, HfLA appreciates your interest in this issue, but please note that it is in the  **"New Issue Approval"**  column of the Project Board because it has not been finalized, approved, or prioritized, and so it is not ready for assignment. For this reason, you have been unassigned from this issue. Please remember to assign issues only from the  **"Prioritized Backlog"**  column.  

The only exceptions to this rule are if you are writing an issue and the `Draft` label is applied, or if you are self-assigning to your "Pre-work Checklist" (the issue includes the `Complexity: Prework` label).

**Comment by HackforLABot** [comment]
Hi @k-cardon, HfLA appreciates your interest in this issue, but please note that it is in the  **"New Issue Approval"**  column of the Project Board because it has not been finalized, approved, or prioritized, and so it is not ready for assignment. For this reason, you have been unassigned from this issue. Please remember to assign issues only from the  **"Prioritized Backlog"**  column.  

The only exceptions to this rule are if you are writing an issue and the `Draft` label is applied, or if you are self-assigning to your "Pre-work Checklist" (the issue includes the `Complexity: Prework` label).

**Comment by HackforLABot** [comment]
Hi @k-cardon, HfLA appreciates your interest in this issue, but please note that it is in the  **"New Issue Approval"**  column of the Project Board because it has not been finalized, approved, or prioritized, and so it is not ready for assignment. For this reason, you have been unassigned from this issue. Please remember to assign issues only from the  **"Prioritized Backlog"**  column.  

The only exceptions to this rule are if you are writing an issue and the `Draft` label is applied, or if you are self-assigning to your "Pre-work Checklist" (the issue includes the `Complexity: Prework` label).

**Comment by k-cardon** [comment]
This issue was in the prioritized backlog, so I assigned myself to it. I was removed because I had another issue-making issue (#6709) for which I was waiting for permission to begin so I could make myself a medium issue. I have now removed myself from the issue making issue so that I could do this medium issue instead; however, it has now been moved out of the "Prioritized Backlog" so that I am not allowed to assign myself anymore.

**Comment by HackforLABot** [comment]
Hi @k-cardon, thank you for taking up this issue! Hfla appreciates you :)

Do let fellow developers know about your:-
i. **Availability:** (When are you available to work on the issue/answer questions other programmers might have about your issue?)
ii. **ETA:** (When do you expect this issue to be completed?)

You're awesome!

P.S. - You may not take up another issue until this issue gets merged (or closed). Thanks again :)

**Comment by k-cardon** [comment]
Availability: 1 hr right now, more time Thursday evening and weekend

Eta: end of week

**Comment by k-cardon** [comment]
Here's a draft of the info for the wiki (I think it would fit on the currently blank page "GitHub Actions" https://hackforla.github.io/website-wiki/roles/dev/github-actions/)

Maintaining GitHub Secrets

The owner of a GitHub repo can create and manage secrets stored by GitHub. It's a good idea to regularly audit and rotate secrets to reduce risk of a compromised secret affecting the product or users. For a guide to implementing, using, and updating secrets stored by GitHub, refer to [this page](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions). 

An audit of secret security would include:

1. Periodically review the registered secrets to confirm they are still required. Remove those that are no longer needed.
2. Reviewing the source code of the repository executing the workflow, and checking any actions used in the workflow. For example, check that they’re not sent to unintended hosts, or explicitly being printed to log output.
3. View the run logs for your workflow after testing valid/invalid inputs, and check that secrets are properly redacted, or not shown.
4. Rotate secrets periodically.

More info [available here](https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-secrets). 

One option is to create some auto-generated issues that pop up periodically, perhaps annually or biannually, that prompt devs to undertake steps 1-3 above, and perhaps whoever owns the repo could undertake step 4 on a similar schedule. The codebase would need to be updated with an issue template for each one and a GitHub action that auto-generates the issue using the template 1-2 times per year.

**Comment by roslynwythe** [comment]
@k-cardon please add to your comments, what is the best way for us to auto-generate issues annually or biannually, to prompt devs to undertake these steps.?  Is that done via a GHA workflow?  Also please add a link to your comment/draft in https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki

**Comment by k-cardon** [comment]
@roslynwythe thanks for your comments! I updated the text and added the link to that contributions page.

**Comment by ExperimentsInHonesty** [comment]
@k-cardon - Please write a decision record (it should include the best practices that you wrote up) drafted in a new comment

Use the template from here https://github.com/hackforla/website/wiki/Decision-Records-on-Solutions-Adopted#template

Once approved, the decision record will end up on these pages
https://github.com/hackforla/website/wiki/Decision-Records-on-Solutions-Adopted
https://github.com/hackforla/website/wiki/Decision-Records

and a link to your final approved draft comment will get linked to this page
https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki

And the decision is to do what you suggest with these specifics
1 GHA to make an issue, bi-annually (twice a year), that has a volunteer with Admin access change the secrets.

**Comment by ExperimentsInHonesty** [comment]
and when you are done creating it, add a `ready for dev lead` label and move it back to Questions / Review column

**Comment by k-cardon** [comment]
This is a record in the [Decision Records on Solutions Adopted](https://github.com/hackforla/website/wiki/Decision-Records-on-Solutions-Adopted).

#### Issue 

Best practices for maintaining GitHub secrets

#### Problem Statement

We store H4LA secrets in GitHub and do not currently have a regular system for maintaining and rotating them, which are best practices for security. 

#### Potential Solution

We could resolve this with: 
1 GHA to make an issue, bi-annually (twice a year), that has a volunteer with Admin access change the secrets, and 
1 GHA to make an issue, bi-annually (twice a year), that prompts a volunteer to audit existing secrets to confirm they are still needed and not unintentionally revealed in logs

#### Feasibility Determination
For a volunteer with Admin access, rotating secrets is a quick and straightforward task. For auditing existing secrets, we will need to establish a list of where and how the secrets are currently used so that it can become a simple task to check the logs and to check whether the secrets are in use. Putting that into place will make both of these tasks routine.

**Comment by roslynwythe** [comment]
Thank you @k-cardon for excellent analysis and writing!
